client/packages/common/src/ui/forms/JsonForms/components/IdGenerator.tsx

```diff
@@ -56,47 +110,95 @@ const validateFields = (
-  data: Record<string, unknown>
-): string => {
-  let output = '';
-  for (const part of options.parts ?? []) {
-    const field = extractField(data, part.field);
-    if (field === undefined || typeof field !== 'string') {
-      continue;
+const mutateString = (value: string, part: StringMutation): string => {
+  if (part.mapping) {
+    const replacement = part.mapping[value];
+    if (replacement !== undefined) {
+      return replacement;
     }
-    let strPart = field;
-    if (part.enumMapping) {
-      const replacement = part.enumMapping[strPart];
-      if (replacement !== undefined) {
-        strPart = replacement;
+  }
+  if (part.firstNChars !== undefined) {
+    return value.slice(0, part.firstNChars);
+  }
+  if (part.lastNChars !== undefined) {
+    return value.slice(value.length - part.lastNChars);
+  }
+  if (part.toUpperCase) {
+    return value.toLocaleUpperCase();
+  }
+  if (part.padString) {
+    return value.padStart(part.padString.length, part.padString);
+  }
+  return value;
+};
+
+const valueFromPart = async (
+  { data, config, allocateNumber }: Omit<GenerateIdInput, 'options'>,
+  part: Part
+): Promise<string | undefined> => {
+  switch (part.type) {
+    case 'Field': {
+      const field = extractField(data, part.field);
+      if (field === undefined || typeof field !== 'string') {
+        return undefined;
       }
+      return field;
     }
-    if (part.firstNChars !== undefined) {
-      strPart = field.slice(0, part.firstNChars);
+    case 'StoreCode': {
+      return config.store?.code;
     }
-    if (part.lastNChars !== undefined) {
-      strPart = field.slice(field.length - part.lastNChars);
+    case 'StoreName': {
+      return config.store?.name;
     }
-    if (part.toUpperCase) {
-      strPart = strPart.toLocaleUpperCase();
+    case 'Number': {
+      return `${await allocateNumber(part.numberName)}`;
     }
+  }
+};
+
+type GenerateIdInput = {
+  options: GeneratorOptions;
+  data: Record<string, unknown>;
+  config: JsonFormsConfig;
+  allocateNumber: (numberName: string) => Promise<number>;
+};
 
-    output += strPart;
+const generateId = async (input: GenerateIdInput): Promise<string> => {
+  let output = '';
+  for (const part of input.options.parts ?? []) {
+    const value = await valueFromPart(input, part);
+    if (value === undefined) {
+      continue;
+    }
+    // apply mutations in sequence
+    output += part.mutations.reduce(
+      (value: string, mutation: StringMutation) =>
+        mutateString(value, mutation),
+      value
+    );
   }
   return output;
 };
 
 const UIComponent = (props: ControlProps) => {
-  const { label, path, data, visible, handleChange, uischema } = props;
+  const { label, path, data, visible, handleChange, uischema, config } = props;
   const t = useTranslation('common');
   const options = uischema.options as GeneratorOptions | undefined;
-  const generate = useCallback(() => {
+  const { mutateAsync: mutateGenerateId } = useMutation(
```

Comment:
:nit - the `mutate` prefix just makes the name confusing to me. Sounds like you are going to mutate the generate result; when I think the operation is simply a `generateId`

```suggestion
  const { mutateAsync: generateId } = useMutation(
```

Reply:
Problem is there is already a generateId method with this name in the file... Maybe generateIdAsync?